### PR TITLE
Imviz: Do not crash on bad GWCS

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to still show the data but without sky coordinates when Glue fails to understand a particular GWCS. In this case, an uncalibrated grism image.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by camipacifici — APPROVED** [review]
Image now loads. Approved

**Comment by pllim** [comment]
I just realized that blink might not work properly when this case is encountered because the current Imviz example notebook depends on successful WCS linking on all data. Also see #667 and #735.

@camipacifici will give this some thoughts, as discussed on Slack.

**Comment by pllim** [comment]
In the mean time, this does solve the use case of displaying _one_ image with bad GWCS.

@camipacifici , do you want to wait or proceed with this PR?

**Comment by camipacifici** [comment]
I think that a user should expect that if an image WCS is not loaded, they will not be able to link it to another image with WCS. I am fine with this PR as is.

**Comment by pllim** [comment]
I am going to close/re-open this PR so the CI kicks off. GitHub Action was down yesterday.

File: jdaviz/configs/imviz/plugins/parsers.py [hunk 1]
```diff
@@ -228,7 +228,13 @@ def _jwst2data(file_obj, ext, data_label):
 
         imdata = getattr(dm, ext)
         component = Component.autotyped(imdata, units=bunit)
-        data.add_component(component=component, label=comp_label)
+
+        # Might have bad GWCS. If so, we exclude it.
+        try:
+            data.add_component(component=component, label=comp_label)
+        except Exception:  # pragma: no cover
```
Comment (pllim) [inline jdaviz/configs/imviz/plugins/parsers.py:235]: I cannot think of a simple way to test this without a lot of mocking or requesting some test data that is not too big and too secretive.

**Review by duytnguyendtn — COMMENTED** [review]
@camipacifici gave me a flat field step file to test this on. I'm getting the following error. Not sure I understand enough to know what this means...
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_7612/771618422.py in <module>
      2 #imviz.load_data(acs_47tuc_1, data_label='acs_47tuc_1')
      3 #imviz.load_data(acs_47tuc_2, data_label='acs_47tuc_2')
----> 4 imviz.load_data(imviz_gwcs, data_label='imviz_gwcs')
      5 viewer = imviz.app.get_viewer('viewer-1')
      6 

\env\lib\site-packages\jdaviz\configs\imviz\helper.py in load_data(self, data, parser_reference, **kwargs)
    102 
    103         else:
--> 104             self.app.load_data(
    105                 data, parser_reference=parser_reference, **kwargs)
    106 

\env\lib\site-packages\jdaviz\app.py in load_data(self, file_obj, parser_reference, **kwargs)
    304                 # If the parser returns something other than known, assume it's
    305                 #  a message we want to make the user aware of.
--> 306                 msg = parser(self, file_obj, **kwargs)
    307 
    308                 if msg is not None:

\env\lib\site-packages\jdaviz\configs\imviz\plugins\parsers.py in parse_data(app, file_obj, ext, data_label, show_in_viewer)
     63         else:  # Assume FITS
     64             with fits.open(file_obj) as pf:
---> 65                 _parse_image(app, pf, data_label, show_in_viewer, ext=ext)
     66     else:
     67         if data_label is None:

\env\lib\site-packages\jdaviz\configs\imviz\plugins\parsers.py in _parse_image(app, file_obj, data_label, show_in_viewer, ext)
    131         raise NotImplementedError(f'Imviz does not support {file_obj}')
    132 
--> 133     for data, data_label in data_iter:
    134 
    135         # avoid duplicate data labels in colection

\env\lib\site-packages\jdaviz\configs\imviz\plugins\parsers.py in _jwst_to_glue_data(file_obj, ext, data_label)
    205             ext = 'data'
    206 
--> 207     data, new_data_label = _jwst2data(file_obj, ext, data_label)
    208 
    209     yield data, new_data_label

\env\lib\site-packages\jdaviz\configs\imviz\plugins\parsers.py in _jwst2data(file_obj, ext, data_label)
    217 
    218     # This is very specific to JWST pipeline image output.
--> 219     with AsdfInFits.open(file_obj) as af:
    220         dm = af.tree
    221         dm_meta = af.tree["meta"]

\env\lib\site-packages\asdf\fits_embed.py in open(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, ignore_missing_extensions, **kwargs)
    216             to be valid.
    217         """
--> 218         return cls._open_impl(fd, uri=uri,
    219                        validate_checksums=validate_checksums,
    220                        extensions=extensions,

\env\lib\site-packages\asdf\fits_embed.py in _open_impl(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, ignore_missing_extensions, **kwargs)
    260 
    261         try:
--> 262             return cls._open_asdf(self, generic_file,
    263                               validate_checksums=validate_checksums,
    264                               extensions=extensions,

\env\lib\site-packages\asdf\asdf.py in _open_asdf(cls, self, fd, validate_checksums, extensions, _get_yaml_content, _force_raw_types, strict_extension_check, ignore_missing_extensions, **kwargs)
    884                     raise
    885 
--> 886             tree = yamlutil.tagged_tree_to_custom_tree(tree, self, _force_raw_types)
    887 
    888             if not (ignore_missing_extensions or _force_raw_types):

\env\lib\site-packages\asdf\yamlutil.py in tagged_tree_to_custom_tree(tree, ctx, force_raw_types, _serialization_context)
    337         return node
    338 
--> 339     return treeutil.walk_and_modify(
    340         tree,
    341         _walker,

\env\lib\site-packages\asdf\treeutil.py in walk_and_modify(top, callback, ignore_implicit_conversion, postorder, _context)
    418 
    419     with _context:
--> 420         return _recurse(top)
    421         # Generators will be drained here, if this is the outermost
    422         # call to walk_and_modify.

\env\lib\site-packages\asdf\treeutil.py in _recurse(node, json_id)
    399                 # If this is a postorder modification, invoke the
    400                 # callback on this node's children first.
--> 401                 result = _handle_children(node, json_id)
    402                 result = _handle_callback(result, json_id)
    403             else:

\env\lib\site-packages\asdf\treeutil.py in _handle_children(node, json_id)
    375             result = node
    376 
--> 377         return _handle_generator(result)
    378 
    379     def _recurse(node, json_id=None):

\env\lib\site-packages\asdf\treeutil.py in _handle_generator(result)
    273         if isinstance(result, types.GeneratorType):
    274             generator = result
--> 275             result = next(generator)
    276             _context.add_generator(generator)
    277 

\env\lib\site-packages\asdf\treeutil.py in _handle_mapping(node, json_id)
    301                 result[key] = PendingValue
    302             else:
--> 303                 value = _recurse(value, json_id)
    304                 if value is not RemoveNode:
    305                     result[key] = value

\env\lib\site-packages\asdf\treeutil.py in _recurse(node, json_id)
    399                 # If this is a postorder modification, invoke the
    400                 # callback on this node's children first.
--> 401                 result = _handle_children(node, json_id)
    402                 result = _handle_callback(result, json_id)
    403             else:

\env\lib\site-packages\asdf\treeutil.py in _handle_children(node, json_id)
    375             result = node
    376 
--> 377         return _handle_generator(result)
    378 
    379     def _recurse(node, json_id=None):

\env\lib\site-packages\asdf\treeutil.py in _handle_generator(result)
    273         if isinstance(result, types.GeneratorType):
    274             generator = result
--> 275             result = next(generator)
    276             _context.add_generator(generator)
    277 

\env\lib\site-packages\asdf\treeutil.py in _handle_mapping(node, json_id)
    301                 result[key] = PendingValue
    302             else:
--> 303                 value = _recurse(value, json_id)
    304                 if value is not RemoveNode:
    305                     result[key] = value

\env\lib\site-packages\asdf\treeutil.py in _recurse(node, json_id)
    400                 # callback on this node's children first.
    401                 result = _handle_children(node, json_id)
--> 402                 result = _handle_callback(result, json_id)
    403             else:
    404                 # Otherwise, invoke the callback on the node first,

\env\lib\site-packages\asdf\treeutil.py in _handle_callback(node, json_id)
    280     def _handle_callback(node, json_id):
    281         if callback_arity == 1:
--> 282             result = callback(node)
    283         else:
    284             result = callback(node, json_id)

\env\lib\site-packages\asdf\yamlutil.py in _walker(node)
    329         # not.
    330         try:
--> 331             return tag_type.from_tree_tagged(node, ctx)
    332         except TypeError as err:
    333             warnings.warn("Failed to convert {} to custom type (detail: {}). "

\env\lib\site-packages\asdf\types.py in from_tree_tagged(cls, tree, ctx)
    358             An instance of the custom type represented by this extension class.
    359         """
--> 360         return cls.from_tree(tree.data, ctx)
    361 
    362     @classmethod

\env\lib\site-packages\gwcs\tags\wcs.py in from_tree(cls, node, ctx)
     27     def from_tree(cls, node, ctx):
     28         name = node['name']
---> 29         steps = [(x.frame, x.transform) for x in node['steps']]
     30         return WCS(steps, name=name)
     31 

\env\lib\site-packages\gwcs\tags\wcs.py in <listcomp>(.0)
     27     def from_tree(cls, node, ctx):
     28         name = node['name']
---> 29         steps = [(x.frame, x.transform) for x in node['steps']]
     30         return WCS(steps, name=name)
     31 

AttributeError: 'TaggedDict' object has no attribute 'frame'
```

**Comment by duytnguyendtn** [comment]
I did merge #778 before reviewing this one. @pllim is this the conflict you were talking about in https://github.com/spacetelescope/jdaviz/pull/726#issuecomment-897130584?

**Comment by pllim** [comment]
Ugh... could it be that grism data needs `jwst` data models after all? But @jdavies-st said we don't need them. Since Imviz cannot read grism GWCS anyway, I can code in an exception handling.

**Comment by pllim** [comment]
I think this works now for Cami's images. Would be happy to add a test if we have a grism image that we can be made public (preferably a small one, like subarray exposure, if possible).

**Comment by camipacifici** [comment]
You can use this one from a JWebbinar: https://stsci.box.com/shared/static/d5k9z5j05dgfv6ljgie483w21kmpevni.fits
It is a little heavy though. I do not have any subarray images handy, sorry.

**Comment by pllim** [comment]
Thanks, Cami! Test added.

**Review by duytnguyendtn — APPROVED** [review]
I was able to load both the test image and the original file @camipacifici gave me! Thanks for the fixes @pllim! Approving.

**Pull request merged**